RFC-002-mapeo-protocol.md

```diff
@@ -0,0 +1,296 @@
+same directory as the upgrades that it serves to others. Downloaded upgrades
+may be placed in a temporary "quarantine" directory first, so that its hash and
+digital signature can be verified, to prevent accidental retransmision of
+broken or fraudulent binaries.
+
+## 4.3 `GetPeerInfo()`
+
+Requests any metadata that the peer wishes to share. At minimum, this should
+include an object of the form
+
+```js
+{
+  plaintextProtocolVersion: "1.0.0",
+  secureProtocolVersion: "1.0.0"
+}
+```
+
+This allows the peer to see if they are able to communicate with the remote
+peer over a secure channel. If their major versions do not match, they can
+perform RPCs like the above to see if one is able to upgrade the other to a
+compatible version.
+
+An implementor should only use the major (`N.x.x`) version number. A minor
+(`x.N.x`) and patch (`x.x.N`) version are included for possible future use.
+
+## 3.4 `Ping()`
+
+A peer should sent periodic `Ping` messages in order to ensure that the
+connection is still active. TCP does not do this automatically, and it can take
+a long time for the kernel to detect that the connection has been lost.
+
+Q: how long between PINGs? how many missed to timeout connection?
+
+## 4.4 `CreateSecureChannel()`
+
+This RPC returns a duplex stream, which first runs a NOISE handshake to
+determine a shared ephemeral key, and then wraps a ChaCha20Poly1305 encryption
+stream around an inner muxrpc stream.
+
+If a secure channel already exists, further calls to this RPC must fail.
+
+# 5. Secure Channel RPCs
+
+The data sync protocol is responsible for letting peers exchange hypercore feed
+entries and media blobs. It happens over the secure channel.
+
+Requests and responses are multiplexed over the [muxrpc][] library, for
+extensibility, and so that several possible subchannels can be established,
+running various protocols:
+
+1. a `hypercore-protocol` stream
+2. a `blob-store-replication-stream` stream
+
+It's important to note that this RPC channel will be running *within* the RPC
+channel from the plaintext channel.
+
+Currently, the RPC protocol is used for coordinating the creation of hypercore
+and blob sync streams. However, it can be used for the exchange of any private
+information.
+
+One such extended use case for the RPC protocol is exchanging metadata about
+what feeds each peer has, how many entries each has, and what blobs are
+present, all before any sync operations occur. This would allow peers to
+generate a diff of what they'd upload and download if they were to sync,
+whether they have enough disk space to do so, etc.
+
+The proposed RPCs are:
+
+## 5.1 `SyncMultifeed()`
+
+This triggers both sides to open a subchannel on the secure multiplexer to hook
+up to the Hypercore sync protocol ([`hypercore-protocol`][]). This results in
+both peers' multifeed databases being consistent with each other.
+
+If a multifeed sync stream is already active, this RPC must fail.
+
+## 5.2 `SyncBlobs()`
```

Comment:
Would it make sense to add an RPC for finding which data the other side has (via bitfield, for example?)

Reply:
Yes, definitely. It's on my internal roadmap, but I decided against further complicating the RFC for now.